src/Microsoft.AspNetCore.Routing/EndpointConstraints/HttpMethodEndpointConstraint.cs

```diff
@@ -55,6 +60,28 @@ public virtual bool Accept(EndpointConstraintContext context)
             var request = context.HttpContext.Request;
             var method = request.Method;
 
+            if (HttpMethodSupported(method))
+            {
+                return true;
+            }
+
+            // Check if request is a CORS OPTIONS request
+            if (string.Equals(request.Method, PreflightHttpMethod, StringComparison.OrdinalIgnoreCase) &&
```

Comment:
Yes, this is what I had in mind, but it also needs to check whether or not the endpoint supports CORS.

Reply:
Hmm, that's a challenge because everything CORS references routing, routing doesn't reference CORS.

What about some CORS metadata interfaces to routing abstractions, have CORS attributes implement interfaces, and MvcEndpointDataSource consume it?

Reply:
I don't think CORS references routing though. 

I'd suggest one modification - CORS interfaces in dispatcher, have MVC detect the attributes (like it does today) and plug in it's own class implementiong of the metadata

Reply:
I checked and it doesn't currently. That's why I thought a routing abstractions reference would be better than a routing reference.

> CORS interfaces in dispatcher

What do you mean by this? Routing package?

Reply:
> What do you mean by this? Routing package?

Reply:
OK I'm dumb https://github.com/aspnet/CORS/blob/master/src/Microsoft.AspNetCore.Cors/Infrastructure/IEnableCorsAttribute.cs

naming isn't great, but CORS already defines something suitable for our purposes. I think it's fine that `Microsoft.AspNetCore.Routing` would reference CORS.

Reply:
For perf the constraint shouldn't be checking the endpoint metadata inside the constraint when it is evaluated, the `MvcEndpointDataSource` should figure that out at startup and set some config on the constraint to tell it that CORS is enabled.

The issue with figuring it out in the endpoint data source is the duplication of a lot of this code which involves looking at attributes to get `IEnableCorsAttribute`: 
https://github.com/aspnet/Mvc/blob/release/2.2/src/Microsoft.AspNetCore.Mvc.Cors/Internal/CorsApplicationModelProvider.cs

Reply:
Do we need to duplicate the code? what if we inspected the results on `ActionDescriptor.Filters`?

Reply:
The filters that `CorsApplicationModelProvider.cs` adds are located in `Microsoft.AspNetCore.Mvc.Cors`, plus they are pub internal.